src/aio.rs

```diff
@@ -32,15 +40,159 @@ use crate::types::{ErrorKind, RedisError, RedisFuture, RedisResult, Value};
+        ) -> core::task::Poll<Result<usize, tokio::io::Error>> {
+            async_std::io::Read::poll_read(Pin::new(&mut self.0), cx, buf)
+        }
+    }
+
+    #[cfg(unix)]
+    impl AsyncWrite for UnixStreamAsyncStdWrapped {
+        fn poll_write(
+            mut self: Pin<&mut Self>,
+            cx: &mut core::task::Context,
+            buf: &[u8],
+        ) -> core::task::Poll<Result<usize, tokio::io::Error>> {
+            async_std::io::Write::poll_write(Pin::new(&mut self.0), cx, buf)
+        }
+        fn poll_flush(
+            mut self: Pin<&mut Self>,
+            cx: &mut core::task::Context,
+        ) -> core::task::Poll<Result<(), tokio::io::Error>> {
+            async_std::io::Write::poll_flush(Pin::new(&mut self.0), cx)
+        }
+
+        fn poll_shutdown(
+            mut self: Pin<&mut Self>,
+            cx: &mut core::task::Context,
+        ) -> core::task::Poll<Result<(), tokio::io::Error>> {
+            async_std::io::Write::poll_close(Pin::new(&mut self.0), cx)
+        }
+    }
+
+    #[cfg(unix)]
+    impl AsyncRead for UnixStreamAsyncStdWrapped {
+        fn poll_read(
+            mut self: Pin<&mut Self>,
+            cx: &mut core::task::Context,
+            buf: &mut [u8],
+        ) -> core::task::Poll<Result<usize, tokio::io::Error>> {
+            async_std::io::Read::poll_read(Pin::new(&mut self.0), cx, buf)
+        }
+    }
+
+    pub struct AsyncStd;
+
+    #[async_trait]
+    impl Connect for AsyncStd {
+        async fn connect_tcp(socket_addr: SocketAddr) -> RedisResult<ActualConnection> {
+            Ok(TcpStreamAsyncStd::connect(&socket_addr)
+                .await
+                .map(|con| ActualConnection::TcpAsyncStd(TcpStreamAsyncStdWrapped(con)))?)
+        }
+        #[cfg(unix)]
+        async fn connect_unix(path: &Path) -> RedisResult<ActualConnection> {
+            Ok(UnixStreamAsyncStd::connect(path)
+                .await
+                .map(|con| ActualConnection::UnixAsyncStd(UnixStreamAsyncStdWrapped(con)))?)
+        }
+    }
+}
+
+#[cfg(feature = "tokio-comp")]
+mod tokio_aio {
+    use super::{async_trait, ActualConnection, Connect, RedisResult, SocketAddr, TcpStreamTokio};
+
     #[cfg(unix)]
-    Unix(Buffered<UnixStream>),
+    use super::{Path, UnixStreamTokio};
+
+    pub struct Tokio;
+
+    #[async_trait]
+    impl Connect for Tokio {
+        async fn connect_tcp(socket_addr: SocketAddr) -> RedisResult<ActualConnection> {
+            Ok(TcpStreamTokio::connect(&socket_addr)
+                .await
+                .map(ActualConnection::TcpTokio)?)
+        }
+        #[cfg(unix)]
+        async fn connect_unix(path: &Path) -> RedisResult<ActualConnection> {
+            Ok(UnixStreamTokio::connect(path)
+                .await
+                .map(ActualConnection::UnixTokio)?)
+        }
+    }
 }
 
-type Buffered<T> = BufReader<BufWriter<T>>;
+/// Represents an async Connection (TCP or Unix. Tokio or Async Std)
+pub enum ActualConnection {
```

Comment:
Does this really need to be public? It becomes part of our public API, which means yet another place that changes depending on features plus the not-so-nice naming.

Reply:
Yeah, because the async_std is in another file now.

Reply:
but can it be `pub(crate)` then?

Reply:
True! I don't know why I didn't think about that :man_facepalming: